src/storage/src/source/persist_source.rs

```diff
@@ -87,6 +87,8 @@ where
             .await
             .expect("could not open persist shard");
 
+        let as_of = as_of.unwrap_or_else(|| read.since().clone());
```

Comment:
Here we could also select `max(since, upper - 1)` instead. This would remove historical detail and make reading the source snapshot more efficient. I just wasn't sure if that may be surprising. Do callers that don't specify an `as_of` expect to get as much historical detail as possible or do they have no expectations about the past

Maybe we should make that decision explicit by introducing an `enum SourceAsOf { At(Antichain<Timestamp>), Earliest, Latest }`.